README.md
# Sofa Http Service [![Build Status](https://travis-ci.org/sofa/sofa-http-service.png?branch=master)](https://travis-ci.org/sofa/sofa-http-service)

This is the Sofa Http Service component. Basically a rip off of angular's `$http` service.

## Installation

Via npm:

```sh
$ npm install sofa-http-service
```

Via Bower:

```sh
$ bower install sofa-http-service
```


CHANGELOG.md
<a name="0.6.0"></a>
## 0.6.0 (2014-12-09)


<a name="0.5.0"></a>
## 0.5.0 (2014-08-05)


<a name="0.4.0"></a>
## 0.4.0 (2014-06-24)


#### Bug Fixes

* **gitignore:** add bower_components ([cbf8eaf2](https://github.com/sofa/sofa-http-service/commit/cbf8eaf2bc3178d6b535ed9e4396c7e7a945df83))


#### Features

* **grunt:** better deploying ([4f515aab](https://github.com/sofa/sofa-http-service/commit/4f515aab0e09e154112ccdb1c6c6c92924306bb1))


<a name="0.3.1"></a>
### 0.3.1 (2014-04-16)


#### Bug Fixes

* **LICENSE:** fixes naming ([28e3c044](https://github.com/sofa/sofa-http-service/commit/28e3c044f9604ecc769929275c75d0bef5c21fe9))
* **karma:** adds sofa-q-service to karma conf ([ee09491e](https://github.com/sofa/sofa-http-service/commit/ee09491e1de7bb246733a3b1eee08618e93f522b))


<a name="0.3.0"></a>
## 0.3.0 (2014-03-27)


#### Features

* **sofa.HttpService:** adds post support ([7ba7f9b9](https://github.com/sofa/sofa-http-service/commit/7ba7f9b95f4a35b521c2c5e9768ec8bee9c755b1))


<a name="0.2.0"></a>
## 0.2.0 (2014-03-26)


#### Features

* **sofa.HttpService:** adds jsonp support ([f3278d7f](https://github.com/sofa/sofa-http-service/commit/f3278d7f2b76e139e19151ec572fca438ec0163e))



test/sofa.httpService.spec.js
'use strict';
/* global sofa */

describe('sofa.httpService', function () {

    var httpService, q;

    beforeEach(function () {
        q = new sofa.QService();
        httpService = new sofa.HttpService(q);
    });

    it('should be defined', function () {
        expect(httpService).toBeDefined();
    });
});


Gruntfile.js
'use strict';
/* jshint camelcase: false */

module.exports = function (grunt) {

    require('load-grunt-tasks')(grunt);

    grunt.initConfig({

        pkg: grunt.file.readJSON('package.json'),
        meta: {
            banner:
                '/**\n' +
                ' * <%= pkg.name %> - v<%= pkg.version %> - <%= grunt.template.today("yyyy-mm-dd") %>\n' +
                ' * <%= pkg.homepage %>\n' +
                ' *\n' +
                ' * Copyright (c) 2014 CouchCommerce GmbH (http://www.couchcommerce.com / http://www.sofa.io) and other contributors\n' +
                ' * THIS SOFTWARE CONTAINS COMPONENTS OF THE SOFA.IO COUCHCOMMERCE SDK (WWW.SOFA.IO).\n' +
                ' * IT IS PROVIDED UNDER THE LICENSE TERMS OF THE ATTACHED LICENSE.TXT.\n' +
                ' */\n'
        },
        component_name: 'sofa.httpService',
        component_sass_name: grunt.file.readJSON('bower.json').name,
        build_dir: 'dist',

        clean: {
            build: {
                src: '<%= build_dir %>'
            }
        },

        html2js: {
            src: {
                options: {
                    base: 'src'
                },
                src: ['src/**/*.tpl.html'],
                dest: '<%= build_dir %>/<%= component_name %>.templates.js',
                module: '<%= component_name %>.templates'
            }
        },

        jshint: {
            options: {
                jshintrc: '.jshintrc'
            },

            src: {
                files: {
                    src: ['src/**/*.js']
                }
            },

            test_unit: {
                files: {
                    src: ['test/**/*.spec.js']
                }
            },

            gruntfile: {
                files: {
                    src: ['Gruntfile.js']
                }
            }
        },

        sass: {
            build: {
                options: {
                    style: 'expanded'
                },
                files: {
                    '<%= build_dir %>/<%= component_sass_name %>.css': 'src/<%= component_sass_name %>.scss',
                    '<%= build_dir %>/<%= component_sass_name %>-default.css': 'src/<%= component_sass_name %>-default.scss'
                }
            }
        },

        copy: {
            build: {
                files: [
                    {
                        src: ['**/*.js'],
                        dest: '<%= build_dir %>/',
                        cwd: 'src',
                        expand: true
                    }
                ]
            }
        },

        karma: {
            options: {
                configFile: 'karma.conf.js'
            },
            unit: {
                port: 9101,
                background: true
            },
            continuous: {
                singleRun: true
            }
        },

        cssmin: {
            compile: {
                options: {
                    report: 'gzip'
                },
                files: {
                    '<%= build_dir %>/<%= component_sass_name %>.css': 'src/<%= component_sass_name %>.css',
                    '<%= build_dir %>/<%= component_sass_name %>-default.css': 'src/<%= component_sass_name %>-default.css'
                }
            }
        },

        ngmin: {
            compile: {
                files: [
                    {
                        src: ['**/*.js'],
                        cwd: '<%= build_dir %>',
                        dest: '<%= build_dir %>',
                        expand: true
                    }
                ]
            }
        },

        concat: {
            options: {
                banner: '<%= meta.banner %>'
            },
            compile_js: {
                src: [
                    'component.prefix',
                    'src/sofa.js',
                    'src/**/*.js',
                    '<%= html2js.src.dest %>',
                    'component.suffix'
                ],
                dest: '<%= build_dir %>/<%= component_name %>.js'
            },
        },

        uglify: {
            compile: {
                options: {
                    banner: '<%= meta.banner %>'
                },
                files: {
                    '<%= build_dir %>/<%= component_name %>.min.js': '<%= concat.compile_js.dest %>'
                }
            },
        },

        changelog: {
            options: {
                dest: 'CHANGELOG.md'
            }
        },

        delta: {

            gruntfile: {
                files: 'Gruntfile.js',
                tasks: ['jshint:gruntfile'],
                options: {
                    livereload: false
                }
            },

            jssrc: {
                files: [
                    'src/**/*.js'
                ],
                tasks: ['jshint:src', 'karma:unit:run', 'ngmin', 'concat:compile_js', 'uglify:compile']
            },

            tpls: {
                files: [
                    '<%= html2js.src.src %>'
                ],
                tasks: ['html2js', 'concat:compile_js', 'uglify:compile']
            },

            jsunit: {
                files: [
                    'test/**/*.spec.js'
                ],
                tasks: ['jshint:test_unit', 'karma:unit:run']
            }
        },

        deploy: {
            options: {
                versionFiles: [
                    'package.json',
                    'bower.json'
                ],
                preDeployFn: function (grunt, newVersion, done) {
                    // Needed to make uglify use the banner with the new version inside
                    grunt.config.set('pkg', grunt.file.readJSON('package.json'));

                    grunt.task.run([
                        'build',
                        'changelog'
                    ]);
                    done();
                },
                postDeployFn: function (grunt, newVersion, done) {
                    if (!grunt.option('soft')) {
                        var exec = require('child_process').exec;
                        exec('npm publish', null, function (err, stdout) {
                            console.log(stdout);
                            done();
                        });
                    }
                    else {
                        done();
                    }
                }
            }
        }
    });

    grunt.registerTask('default', 'build');
    grunt.renameTask('watch', 'delta');

    grunt.registerTask('watch', [
        'build',
        'karma:unit',
        'delta'
    ]);

    grunt.registerTask('test', [
        'jshint',
        'karma:continuous'
    ]);

    grunt.registerTask('build', [
        'clean',
        'jshint',
        'sass',
        'karma:continuous',
        'cssmin:compile',
        'ngmin',
        'concat:compile_js',
        'uglify:compile'
    ]);
};


karma.conf.js
module.exports = function (karma) {
    karma.set({
        /**
         * From where to look for files, starting with the location of this file.
         */
        basePath: '.',

        /**
         * This is the list of file patterns to load into the browser during testing.
         */
        files: [
            'node_modules/sofa-core/dist/sofa.core.js',
            'src/**/*.js',
            'node_modules/sofa-testing/mocks/sofa.config.mock.js',
            'node_modules/sofa-q-service/dist/sofa.qService.js',
            'test/**/*.spec.js'
        ],
        exclude: [],
        frameworks: ['jasmine'],
        plugins: [
            'karma-jasmine',
            'karma-firefox-launcher',
            'karma-chrome-launcher'
        ],
        preprocessors: {},

        /**
         * How to report, by default.
         */
        reporters: 'dots',

        /**
         * On which port should the browser connect, on which port is the test runner
         * operating, and what is the URL path for the browser to use.
         */
        port: 9018,
        runnerPort: 9101,
        urlRoot: '/',

        /**
         * Disable file watching by default.
         */
        autoWatch: false,

        /**
         * The list of browsers to launch to test on. This includes only "Firefox" by
         * default, but other browser names include:
         * Chrome, ChromeCanary, Firefox, Opera, Safari, PhantomJS
         *
         * Note that you can also use the executable name of the browser, like "chromium"
         * or "firefox", but that these vary based on your operating system.
         *
         * You may also leave this blank and manually navigate your browser to
         * http://localhost:9018/ when you're running tests. The window/tab can be left
         * open and the tests will automatically occur there during the build. This has
         * the aesthetic advantage of not launching a browser every time you save.
         */
        browsers: [
            process.env.TRAVIS ? 'Firefox' : 'Chrome'
        ]
    });
};


src/sofa.httpService.js
'use strict';
/* global sofa */
/* global XMLHttpRequest */

/**
 * @sofadoc class
 * @name sofa.HttpService
 * @package sofa-http-service
 * @requiresPackage sofa-q-service
 * @requires sofa.QService
 * @distFile dist/sofa.HttpService
 *
 * @description
 */
sofa.define('sofa.HttpService', function ($q) {

    var JSON_START = /^\s*(\[|\{[^\{])/,
        JSON_END = /[\}\]]\s*$/,
        PROTECTION_PREFIX = /^\)\]\}',?\n/,
        CONTENT_TYPE_APPLICATION_JSON = {
            'Content-Type': 'application/json;charset=utf-8'
        },
        ABORTED = -1;

    var rawDocument = window.document;

    function encodeUriQuery(val, pctEncodeSpaces) {
        return encodeURIComponent(val)
                    .replace(/%40/gi, '@')
                    .replace(/%3A/gi, ':')
                    .replace(/%24/g, '$')
                    .replace(/%2C/gi, ',')
                    .replace(/%20/g, (pctEncodeSpaces ? '%20' : '+'));
    }

    function indexOf(array, obj) {
        if (array.indexOf) {
            return array.indexOf(obj);
        }

        for (var i = 0; i < array.length; i++) {
            if (obj === array[i]) {
                return i;
            }
        }
        return -1;
    }

    var isString = sofa.Util.isString,
        toJson = sofa.Util.toJson,
        isObject = sofa.Util.isObject,
        isFunction = sofa.Util.isFunction,
        isUndefined = sofa.Util.isUndefined,
        isDefined = function (value) {
            return !isUndefined(value);
        },
        isArray = sofa.Util.isArray,
        forEach = sofa.Util.forEach,
        extend = sofa.Util.extend;


    var fromJson = function (json) {
        return sofa.Util.isString(json) ? JSON.parse(json) : json;
    };

    var lowercase = function (string) {
        return isString(string) ? string.toLowerCase() : string;
    };

    var uppercase = function (string) {
        return isString(string) ? string.toUpperCase() : string;
    };

    var trim = function (value) {
        return isString(value) ? value.trim() : value;
    };

    function transformData(data, headers, fns) {
        if (isFunction(fns)) {
            return fns(data, headers);
        }

        forEach(fns, function (fn) {
            data = fn(data, headers);
        });

        return data;
    }

    function isSuccess(status) {
        return 200 <= status && status < 300;
    }

    function createXhr() {
        return new XMLHttpRequest();
    }

    function headersGetter(headers) {
        var headersObj = isObject(headers) ? headers : undefined;

        return function (name) {
            if (!headersObj) {
                headersObj =  parseHeaders(headers);
            }

            if (name) {
                return headersObj[lowercase(name)] || null;
            }

            return headersObj;
        };
    }

    function parseHeaders(headers) {
        var parsed = {}, key, val, i;

        if (!headers) {
            return parsed;
        }

        forEach(headers.split('\n'), function (line) {
            i = line.indexOf(':');
            key = lowercase(trim(line.substr(0, i)));
            val = trim(line.substr(i + 1));

            if (key) {
                if (parsed[key]) {
                    parsed[key] += ', ' + val;
                } else {
                    parsed[key] = val;
                }
            }
        });

        return parsed;
    }

    var defaults = this.defaults = {
        // transform incoming response data
        transformResponse: [function (data) {
            if (isString(data)) {
                // strip json vulnerability protection prefix
                data = data.replace(PROTECTION_PREFIX, '');
                if (JSON_START.test(data) && JSON_END.test(data)) {
                    data = fromJson(data);
                }
            }
            return data;
        }],

        // transform outgoing request data
        transformRequest: [function (d) {
            return isObject(d) ? toJson(d) : d;
        }],

        // default headers
        headers: {
            common: {
                'Accept': 'application/json, text/plain, */*'
            },
            post:   sofa.Util.clone(CONTENT_TYPE_APPLICATION_JSON),
            put:    sofa.Util.clone(CONTENT_TYPE_APPLICATION_JSON),
            patch:  sofa.Util.clone(CONTENT_TYPE_APPLICATION_JSON)
        },

        xsrfCookieName: 'XSRF-TOKEN',
        xsrfHeaderName: 'X-XSRF-TOKEN'
    };

    var httpService = function (requestConfig) {

        var config = {
            method: 'get',
            transformRequest: defaults.transformRequest,
            transformResponse: defaults.transformResponse
        };

        var headers = mergeHeaders(requestConfig);

        extend(config, requestConfig);
        config.headers = headers;
        config.method = uppercase(config.method);

        var serverRequest = function (config) {
            headers = config.headers;
            var reqData = transformData(config.data, headersGetter(headers), config.transformRequest);

            // strip content-type if data is undefined
            if (isUndefined(config.data)) {
                forEach(headers, function (value, header) {
                    if (lowercase(header) === 'content-type') {
                        delete headers[header];
                    }
                });
            }

            if (isUndefined(config.withCredentials) && !isUndefined(defaults.withCredentials)) {
                config.withCredentials = defaults.withCredentials;
            }

            // send request
            return sendReq(config, reqData, headers).then(transformResponse, transformResponse);
        };

        var chain = [serverRequest, undefined];
        var promise = $q.when(config);

        while (chain.length) {
            var thenFn = chain.shift();
            var rejectFn = chain.shift();

            promise = promise.then(thenFn, rejectFn);
        }

        return promise;

        function transformResponse(response) {
            // make a copy since the response must be cacheable
            var resp = extend({}, response, {
                data: transformData(response.data, response.headers, config.transformResponse)
            });
            return (isSuccess(response.status)) ? resp : $q.reject(resp);
        }

        function mergeHeaders(config) {
            var defHeaders = defaults.headers,
                reqHeaders = extend({}, config.headers),
                defHeaderName, lowercaseDefHeaderName, reqHeaderName;

            defHeaders = extend({}, defHeaders.common, defHeaders[lowercase(config.method)]);

            // execute if header value is function
            execHeaders(defHeaders);
            execHeaders(reqHeaders);

            // using for-in instead of forEach to avoid unecessary iteration after header has been found
            defaultHeadersIteration:
            for (defHeaderName in defHeaders) {
                lowercaseDefHeaderName = lowercase(defHeaderName);

                for (reqHeaderName in reqHeaders) {
                    if (lowercase(reqHeaderName) === lowercaseDefHeaderName) {
                        continue defaultHeadersIteration;
                    }
                }
                reqHeaders[defHeaderName] = defHeaders[defHeaderName];
            }

            return reqHeaders;

            function execHeaders(headers) {
                var headerContent;

                forEach(headers, function (headerFn, header) {
                    if (isFunction(headerFn)) {
                        headerContent = headerFn();
                        if (headerContent !== null) {
                            headers[header] = headerContent;
                        } else {
                            delete headers[header];
                        }
                    }
                });
            }
        }
    };

    httpService.pendingRequests = [];
    httpService.callbacks = {
        counter: 0
    };
    httpService.defaults = defaults;

    function sortedKeys(obj) {
        var keys = [];
        for (var key in obj) {
            if (obj.hasOwnProperty(key)) {
                keys.push(key);
            }
        }
        return keys.sort();
    }

    function forEachSorted(obj, iterator, context) {
        var keys = sortedKeys(obj);
        for (var i = 0; i < keys.length; i++) {
            iterator.call(context, obj[keys[i]], keys[i]);
        }
        return keys;
    }

    function buildUrl(url, params) {
        if (!params) {
            return url;
        }
        var parts = [];
        forEachSorted(params, function (value, key) {
            if (value === null || isUndefined(value)) {
                return;
            }

            if (!isArray(value)) {
                value = [value];
            }

            forEach(value, function (v) {
                if (isObject(v)) {
                    v = toJson(v);
                }
                parts.push(encodeUriQuery(key) + '=' +
                            encodeUriQuery(v));
            });
        });

        if (parts.length > 0) {
            url += ((url.indexOf('?') === -1) ? '?' : '&') + parts.join('&');
        }
        return url;
    }

    function sendReq(config, reqData, reqHeaders) {
        var deferred = $q.defer(),
            promise = deferred.promise,
            cache,
            cachedResp,
            url = buildUrl(config.url, config.params);

        httpService.pendingRequests.push(config);
        promise.then(removePendingReq, removePendingReq);


        // if we won't have the response in cache, send the request to the backend
        if (isUndefined(cachedResp)) {
            doActualRequest(config.method, url, reqData, done, reqHeaders, config.timeout,
                config.withCredentials, config.responseType);
        }

        return promise;


        /**
        * Callback registered to $httpBackend():
        *  - caches the response if desired
        *  - resolves the raw $http promise
        *  - calls $apply
        */
        function done(status, response, headersString) {
            if (cache) {
                if (isSuccess(status)) {
                    cache.put(url, [status, response, parseHeaders(headersString)]);
                } else {
                    // remove promise from the cache
                    cache.remove(url);
                }
            }

            resolvePromise(response, status, headersString);
        }


        /**
        * Resolves the raw $http promise.
        */
        function resolvePromise(response, status, headers) {
            // normalize internal statuses to 0
            status = Math.max(status, 0);

            (isSuccess(status) ? deferred.resolve : deferred.reject)({
                data: response,
                status: status,
                headers: headersGetter(headers),
                config: config
            });
        }


        function removePendingReq() {
            var idx = indexOf(httpService.pendingRequests, config);
            if (idx !== -1) {
                httpService.pendingRequests.splice(idx, 1);
            }
        }
    }

    function doActualRequest(method, url, post, callback, headers, timeout, withCredentials, responseType) {

        var status;
        url = url;
        var callbacks = httpService.callbacks;

        if (lowercase(method) === 'jsonp') {
            var callbackId = '_' + (callbacks.counter++).toString(36);
            callbacks[callbackId] = function (data) {
                callbacks[callbackId].data = data;
            };

            var jsonpDone = jsonpReq(url.replace('JSON_CALLBACK', 'sofa.callbacks.' + callbackId),  function () {
                if (callbacks[callbackId].data) {
                    completeRequest(callback, 200, callbacks[callbackId].data);
                } else {
                    completeRequest(callback, status || -2);
                }
                callbacks[callbackId] = function () {};
            });
        } else {
            var xhr = createXhr(method);

            xhr.open(method, url, true);
            forEach(headers, function (value, key) {
                if (isDefined(value)) {
                    xhr.setRequestHeader(key, value);
                }
            });

            // In IE6 and 7, this might be called synchronously when xhr.send below is called and the
            // response is in the cache. the promise api will ensure that to the app code the api is
            // always async
            xhr.onreadystatechange = function () {
                // onreadystatechange might get called multiple times with readyState === 4 on mobile webkit caused by
                // xhrs that are resolved while the app is in the background (see #5426).
                // since calling completeRequest sets the `xhr` variable to null, we just check if it's not null before
                // continuing
                //
                // we can't set xhr.onreadystatechange to undefined or delete it because that breaks IE8 (method=PATCH) and
                // Safari respectively.
                if (xhr && xhr.readyState === 4) {
                    var responseHeaders = null,
                        response = null;

                    if (status !== ABORTED) {
                        responseHeaders = xhr.getAllResponseHeaders();

                        // responseText is the old-school way of retrieving response (supported by IE8 & 9)
                        // response/responseType properties were introduced in XHR Level2 spec (supported by IE10)
                        response = ('response' in xhr) ? xhr.response : xhr.responseText;
                    }

                    completeRequest(callback,
                        status || xhr.status,
                        response,
                        responseHeaders);
                }
            };

            if (responseType) {
                try {
                    xhr.responseType = responseType;
                } catch (e) {
                    // WebKit added support for the json responseType value on 09/03/2013
                    // https://bugs.webkit.org/show_bug.cgi?id=73648. Versions of Safari prior to 7 are
                    // known to throw when setting the value "json" as the response type. Other older
                    // browsers implementing the responseType
                    //
                    // The json response type can be ignored if not supported, because JSON payloads are
                    // parsed on the client-side regardless.
                    if (responseType !== 'json') {
                        throw e;
                    }
                }
            }

            xhr.send(post || null);
        }

        function completeRequest(callback, status, response, headersString) {

            jsonpDone = xhr = null;
            // fix status code when it is 0 (0 status is undocumented).
            // Occurs when accessing file resources.
            // On Android 4.1 stock browser it occurs while retrieving files from application cache.
            status = (status === 0) ? (response ? 200 : 404) : status;
            callback(status, response, headersString);
        }

        function jsonpReq(url, done) {
            // we can't use jQuery/jqLite here because jQuery does crazy shit with script elements, e.g.:
            // - fetches local scripts via XHR and evals them
            // - adds and immediately removes script elements from the document
            var script = rawDocument.createElement('script'),
                doneWrapper = function () {
                    script.onreadystatechange = script.onload = script.onerror = null;
                    rawDocument.body.removeChild(script);
                    if (done) {
                        done();
                    }
                };

            script.type = 'text/javascript';
            script.src = url;

            script.onload = script.onerror = function () {
                doneWrapper();
            };

            rawDocument.body.appendChild(script);
            return doneWrapper;
        }
    }

    var createShortMethods = function () {
        forEach(arguments, function (name) {
            httpService[name] = function (url, config) {
                return httpService(extend(config || {}, {
                    method: name,
                    url: url
                }));
            };
        });
    };

    var createShortMethodsWithData = function () {
        forEach(arguments, function (name) {
            httpService[name] = function (url, data, config) {
                return httpService(extend(config || {}, {
                    method: name,
                    url: url,
                    data: data
                }));
            };
        });
    };

    createShortMethods('get', 'jsonp');
    createShortMethodsWithData('post');

    return httpService;
});
